NTupleAnalyzerHH_nanoAOD.py

```diff
@@ -846,45 +899,72 @@ def getLeptonEventFlavorGEN(T):
-
-		if pdg not in [-15,-13,-11,11,13,15]:
-			continue
-		if motherid in [23]:
-		#if motherid in [-23,23]:
-		#	grandmotherid = 0
-		#       grandmotherIndex = T.GenParticleMotherIndex[motherIndex]
-		#	if grandmotherIndex>-1:
-		#       grandmotherid = T.GenParticlePdgId[motherIndex]
-		#		print pdg,motherid,grandmotherid
-		#	if grandmotherid not in [-25,25]: continue
-		#print n,T.GenParticlePdgId[n],T.GenParticleStatus[n],T.GenParticleMass[n],T.GenParticlePdgId[T.GenParticleMotherIndex[n]],grandmotherid,T.GenParticleNumDaught[n],bool(T.GenParticleIsLastCopy[n])
-			if pdg in [-11,11]:
-				isEleEvent=True
-				return [isEleEvent,isMuEvent,isTauEvent]
-			if pdg in [-13,13]:
-				isMuEvent=True
-				return [isEleEvent,isMuEvent,isTauEvent]
-			if pdg in [-15,15]:
-				isTauEvent=True
-				return [isEleEvent,isMuEvent,isTauEvent]
-			#if (isEleEvent or isMuEvent or isTauEvent) : gotIt=True
+
+	if isData==False:
+
+		for n in range(T.nGenPart):
+			#if gotIt==True: continue
+			pdg = T.GenPart_pdgId[n]
+			#if pdg in [-15,-13,-11,11,13,15]:
+			#	print n,T.GenParticlePdgId[n],T.GenParticleStatus[n],T.GenParticleMass[n],T.GenParticlePdgId[T.GenParticleMotherIndex[n]],T.GenParticleNumDaught[n],bool(T.GenParticleIsLastCopy[n])
+			motherIndex = T.GenPart_genPartIdxMother[n]
+
+			if motherIndex >= 0:
+				motherid = T.GenPart_pdgId[motherIndex]
+			#if motherid in [-15,15]:
+			#	print n,T.GenParticlePdgId[n],T.GenParticleStatus[n],T.GenParticleMass[n],T.GenParticlePdgId[T.GenParticleMotherIndex[n]],T.GenParticleNumDaught[n],bool(T.GenParticleIsLastCopy[n])
+
+			if pdg not in [-15,-13,-11,11,13,15]:
+				continue
+			if motherid in [23]:
+			#if motherid in [-23,23]:
+			#	grandmotherid = 0
+			#       grandmotherIndex = T.GenParticleMotherIndex[motherIndex]
+			#	if grandmotherIndex>-1:
+			#       grandmotherid = T.GenParticlePdgId[motherIndex]
+			#		print pdg,motherid,grandmotherid
+			#	if grandmotherid not in [-25,25]: continue
+			#print n,T.GenParticlePdgId[n],T.GenParticleStatus[n],T.GenParticleMass[n],T.GenParticlePdgId[T.GenParticleMotherIndex[n]],grandmotherid,T.GenParticleNumDaught[n],bool(T.GenParticleIsLastCopy[n])
+				if pdg in [-11,11]:
+					isEleEvent=True
+					return [isEleEvent,isMuEvent,isTauEvent]
+				if pdg in [-13,13]:
+					isMuEvent=True
+					return [isEleEvent,isMuEvent,isTauEvent]
+				if pdg in [-15,15]:
+					isTauEvent=True
+					return [isEleEvent,isMuEvent,isTauEvent]
+				#if (isEleEvent or isMuEvent or isTauEvent) : gotIt=True
 	return [isEleEvent,isMuEvent,isTauEvent]
 
-def LeptonsAndJetsFromHH(T):
+def LeptonsAndJetsFromHH(T,radius,isdata):
 	# Purpose: Testing. HH. Get the muons from Z decays (on and off shell) and find the matching reco muons.
 	#         Return TLorentzVectors of the gen and reco muons, and the indices for
 	#         the recomuons as well.
+
+	if radius == 4:
```

Comment:
While it shouldn't happen, let's protect ourselves against human error. Currently only choices are radius=4 or 8. What if someone in the future puts 6? You can make e.g. a last 'elif' which prints an informative error and kills the whole job. As a general point you should always think about all the possible error modes of a function - case in point these acosine functions that were throwing errors when the input was out of range....